File: js/services/quotation-service.js
import { database } from '../firebase-config.js';
import { ref, push, set, get, remove, query, orderByChild, limitToLast } from 'firebase/database';

export class QuotationService {
    constructor() {
        this.path = 'quotations';
    }

    async create(quotation) {
        const quotationRef = ref(database, this.path);
        const newRef = push(quotationRef);
        await set(newRef, quotation);
        return newRef.key;
    }

    async update(id, quotation) {
        const quotationRef = ref(database, `${this.path}/${id}`);
        await set(quotationRef, quotation);
    }

    async delete(id) {
        const quotationRef = ref(database, `${this.path}/${id}`);
        await remove(quotationRef);
    }

    async get(id) {
        const quotationRef = ref(database, `${this.path}/${id}`);
        const snapshot = await get(quotationRef);
        if (snapshot.exists()) {
            return { id: snapshot.key, ...snapshot.val() };
        }
        return null;
    }

    async getAll() {
        const quotationsRef = ref(database, this.path);
        const snapshot = await get(quotationsRef);
        const quotations = [];
        if (snapshot.exists()) {
            snapshot.forEach(child => {
                quotations.push({ id: child.key, ...child.val() });
            });
        }
        return quotations;
    }

    async generateQuotationNumber() {
        const quotationsRef = ref(database, this.path);
        const q = query(quotationsRef, orderByChild('quotationNumber'), limitToLast(1));
        const snapshot = await get(q);
        
        let nextNumber = 1;
        if (snapshot.exists()) {
            snapshot.forEach(child => {
                const quotation = child.val();
                if (quotation.quotationNumber) {
                    const match = quotation.quotationNumber.match(/\d+/);
                    if (match) {
                        nextNumber = parseInt(match[0]) + 1;
                    }
                }
            });
        }
        
        return `QUO-${String(nextNumber).padStart(4, '0')}`;
    }
}

export const quotationService = new QuotationService();

File: lib/services/quotation_service.dart
import '../models/quotation.dart';
import 'firebase_service.dart';

class QuotationService {
  final FirebaseService _firebase;
  final String _path = 'quotations';

  QuotationService(this._firebase);

  Future<String> create(Quotation quotation) async {
    return await _firebase.save(_path, quotation.toJson());
  }

  Future<void> update(Quotation quotation) async {
    if (quotation.id == null) throw Exception('Quotation ID is required');
    await _firebase.update(_path, quotation.id!, quotation.toJson());
  }

  Future<void> delete(String id) async {
    await _firebase.delete(_path, id);
  }

  Future<Quotation?> get(String id) async {
    final data = await _firebase.get(_path, id);
    return data != null ? Quotation.fromJson(data, id) : null;
  }

  Future<List<Quotation>> getAll() async {
    final data = await _firebase.getAll(_path);
    return data.map((item) => Quotation.fromJson(item, item['id'])).toList();
  }

  Future<String> generateQuotationNumber() async {
    final quotations = await getAll();
    int nextNumber = 1;
    
    if (quotations.isNotEmpty) {
      final numbers = quotations
          .map((quo) {
            final match = RegExp(r'\d+').firstMatch(quo.quotationNumber);
            return match != null ? int.tryParse(match.group(0)!) : 0;
          })
          .where((n) => n != null && n > 0)
          .cast<int>() // Explicit cast to ensure type safety
          .toList();
      
      if (numbers.isNotEmpty) {
        nextNumber = numbers.reduce((a, b) => a > b ? a : b) + 1;
      }
    }
    
    return 'QUO-${nextNumber.toString().padLeft(4, '0')}';
  }
}


